File: mem0/vector_stores/milvus.py
import logging
from typing import Dict, Optional

from pydantic import BaseModel

from mem0.configs.vector_stores.milvus import MetricType
from mem0.vector_stores.base import VectorStoreBase

try:
    import pymilvus  # noqa: F401
except ImportError:
    raise ImportError("The 'pymilvus' library is required. Please install it using 'pip install pymilvus'.")

from pymilvus import CollectionSchema, DataType, FieldSchema, Function, FunctionType, MilvusClient

logger = logging.getLogger(__name__)


class OutputData(BaseModel):
    id: Optional[str]  # memory id
    score: Optional[float]  # distance
    payload: Optional[Dict]  # metadata


class MilvusDB(VectorStoreBase):
    def __init__(
        self,
        url: str,
        token: str,
        collection_name: str,
        embedding_model_dims: int,
        metric_type: MetricType,
        db_name: str,
    ) -> None:
        """Initialize the MilvusDB database.

        Args:
            url (str): Full URL for Milvus/Zilliz server.
            token (str): Token/api_key for Zilliz server / for local setup defaults to None.
            collection_name (str): Name of the collection (defaults to mem0).
            embedding_model_dims (int): Dimensions of the embedding model (defaults to 1536).
            metric_type (MetricType): Metric type for similarity search (defaults to L2).
            db_name (str): Name of the database (defaults to "").
        """
        self.collection_name = collection_name
        self.embedding_model_dims = embedding_model_dims
        self.metric_type = metric_type
        self.client = MilvusClient(uri=url, token=token, db_name=db_name)
        # Whether this collection has the `text` + `sparse` fields for v3 BM25.
        # Pre-v3 collections lack them; writing a top-level `text` field is rejected.
        self._has_bm25_schema = False
        self.create_col(
            collection_name=self.collection_name,
            vector_size=self.embedding_model_dims,
            metric_type=self.metric_type,
        )

    def create_col(
        self,
        collection_name: str,
        vector_size: int,
        metric_type: MetricType = MetricType.COSINE,
    ) -> None:
        """Create a new collection with index_type AUTOINDEX.

        Args:
            collection_name (str): Name of the collection (defaults to mem0).
            vector_size (int): Dimensions of the embedding model (defaults to 1536).
            metric_type (MetricType, optional): etric type for similarity search. Defaults to MetricType.COSINE.
        """

        if self.client.has_collection(collection_name):
            logger.info(f"Collection {collection_name} already exists. Skipping creation.")
            desc = self.client.describe_collection(collection_name=collection_name)
            field_names = {f.get("name") for f in desc.get("fields", [])}
            self._has_bm25_schema = "text" in field_names and "sparse" in field_names
            if not self._has_bm25_schema:
                logger.warning(
                    f"Collection '{collection_name}' predates v3 hybrid search (no 'text'/'sparse' fields). "
                    "BM25 keyword scoring will be disabled for this collection; semantic search works normally. "
                    "To enable hybrid search, use a fresh collection."
                )
        else:
            fields = [
                FieldSchema(name="id", dtype=DataType.VARCHAR, is_primary=True, max_length=512),
                FieldSchema(name="vectors", dtype=DataType.FLOAT_VECTOR, dim=vector_size),
                FieldSchema(name="metadata", dtype=DataType.JSON),
                # Text field for BM25 full-text search (auto-tokenized by Milvus analyzer)
                FieldSchema(name="text", dtype=DataType.VARCHAR, max_length=65535, enable_analyzer=True),
                # Sparse vector field populated automatically by the BM25 function below
                FieldSchema(name="sparse", dtype=DataType.SPARSE_FLOAT_VECTOR),
            ]

            schema = CollectionSchema(fields, enable_dynamic_field=True)

            # Add BM25 function so Milvus auto-generates sparse vectors from the text field
            bm25_function = Function(
                name="bm25",
                input_field_names=["text"],
                output_field_names=["sparse"],
                function_type=FunctionType.BM25,
            )
            schema.add_function(bm25_function)

            index_params = self.client.prepare_index_params()
            index_params.add_index(
                field_name="vectors", metric_type=metric_type, index_type="AUTOINDEX", index_name="vector_index"
            )
            index_params.add_index(
                field_name="sparse",
                index_type="SPARSE_INVERTED_INDEX",
                metric_type="BM25",
                index_name="sparse_index",
            )
            self.client.create_collection(collection_name=collection_name, schema=schema, index_params=index_params)
            self._has_bm25_schema = True

    def insert(self, ids, vectors, payloads, **kwargs: Optional[dict[str, any]]):
        """Insert vectors into a collection.

        Args:
            vectors (List[List[float]]): List of vectors to insert.
            payloads (List[Dict], optional): List of payloads corresponding to vectors.
            ids (List[str], optional): List of IDs corresponding to vectors.
        """
        # Batch insert all records at once for better performance and consistency.
        # Only include the `text` field when the collection's schema has it — legacy
        # collections created pre-v3 reject unknown top-level fields.
        def _build_record(idx, embedding, metadata):
            record = {"id": idx, "vectors": embedding, "metadata": metadata}
            if self._has_bm25_schema:
                # Populate the text field for BM25 sparse search; prefer lemmatized text, fall back to raw data
                record["text"] = (metadata.get("text_lemmatized") or metadata.get("data", ""))[:65535] if metadata else ""
            return record

        data = [_build_record(idx, embedding, metadata) for idx, embedding, metadata in zip(ids, vectors, payloads)]
        self.client.insert(collection_name=self.collection_name, data=data, **kwargs)

    def _create_filter(self, filters: dict):
        """Prepare filters for efficient query.

        Args:
            filters (dict): filters [user_id, agent_id, run_id]

        Returns:
            str: formated filter.
        """
        operands = []
        for key, value in filters.items():
            if isinstance(value, str):
                operands.append(f'(metadata["{key}"] == "{value}")')
            else:
                operands.append(f'(metadata["{key}"] == {value})')

        return " and ".join(operands)

    def _parse_output(self, data: list):
        """
        Parse the output data.

        Args:
            data (Dict): Output data.

        Returns:
            List[OutputData]: Parsed output data.
        """
        memory = []

        for value in data:
            uid, score, metadata = (
                value.get("id"),
                value.get("distance"),
                value.get("entity", {}).get("metadata"),
            )

            memory_obj = OutputData(id=uid, score=score, payload=metadata)
            memory.append(memory_obj)

        return memory

    def search(self, query: str, vectors: list, top_k: int = 5, filters: dict = None) -> list:
        """
        Search for similar vectors.

        Args:
            query (str): Query.
            vectors (List[float]): Query vector.
            top_k (int, optional): Number of results to return. Defaults to 5.
            filters (Dict, optional): Filters to apply to the search. Defaults to None.

        Returns:
            list: Search results.
        """
        query_filter = self._create_filter(filters) if filters else None
        # v3 collections carry both a dense `vectors` field and a sparse `sparse`
        # field (for BM25), which makes anns_field ambiguous — Milvus rejects the
        # query otherwise with "multiple anns_fields exist". Legacy single-vector
        # collections don't need the hint, so only pass it when the hybrid schema
        # is present.
        search_kwargs = {
            "collection_name": self.collection_name,
            "data": [vectors],
            "limit": top_k,
            "filter": query_filter,
            "output_fields": ["*"],
        }
        if self._has_bm25_schema:
            search_kwargs["anns_field"] = "vectors"
        hits = self.client.search(**search_kwargs)
        result = self._parse_output(data=hits[0])
        return result

    def keyword_search(self, query, top_k=5, filters=None):
        """
        Search for memories using BM25-based full-text search via Milvus sparse vector support.

        Milvus 2.5+ supports native BM25 via full-text search with a SPARSE_FLOAT_VECTOR field.
        This method attempts to use that capability. If the collection does not have a sparse
        field configured, it returns None gracefully.

        Args:
            query (str): The text query for keyword-based search.
            top_k (int, optional): Number of results to return. Defaults to 5.
            filters (dict, optional): Filters to apply to the search. Defaults to None.

        Returns:
            list: Search results in the same format as search(), or None if sparse search
                  is not supported on this collection.
        """
        if not self._has_bm25_schema:
            return None
        try:
            query_filter = self._create_filter(filters) if filters else None
            hits = self.client.search(
                collection_name=self.collection_name,
                data=[query],
                anns_field="sparse",
                limit=top_k,
                filter=query_filter,
                output_fields=["*"],
            )
            result = self._parse_output(data=hits[0])
            return result
        except Exception as e:
            logger.debug(f"Keyword search not available for collection {self.collection_name}: {e}")
            return None

    def delete(self, vector_id):
        """
        Delete a vector by ID.

        Args:
            vector_id (str): ID of the vector to delete.
        """
        self.client.delete(collection_name=self.collection_name, ids=vector_id)

    def update(self, vector_id=None, vector=None, payload=None):
        """
        Update a vector and its payload.

        Args:
            vector_id (str): ID of the vector to update.
            vector (List[float], optional): Updated vector.
            payload (Dict, optional): Updated payload.
        """
        if vector is None or payload is None:
            existing = self.client.get(collection_name=self.collection_name, ids=vector_id)
            if not existing:
                raise ValueError(f"Vector with id {vector_id} not found in collection {self.collection_name}")
            if vector is None:
                vector = existing[0].get("vectors")
                if vector is None:
                    raise ValueError(f"Existing record {vector_id} has no vector data")
            if payload is None:
                payload = existing[0].get("metadata")

        text = ""
        if payload:
            text = (payload.get("text_lemmatized") or payload.get("data", ""))[:65535]
        schema = {"id": vector_id, "vectors": vector, "metadata": payload, "text": text}
        self.client.upsert(collection_name=self.collection_name, data=schema)

    def get(self, vector_id):
        """
        Retrieve a vector by ID.

        Args:
            vector_id (str): ID of the vector to retrieve.

        Returns:
            OutputData: Retrieved vector.
        """
        result = self.client.get(collection_name=self.collection_name, ids=vector_id)
        output = OutputData(
            id=result[0].get("id", None),
            score=None,
            payload=result[0].get("metadata", None),
        )
        return output

    def list_cols(self):
        """
        List all collections.

        Returns:
            List[str]: List of collection names.
        """
        return self.client.list_collections()

    def delete_col(self):
        """Delete a collection."""
        return self.client.drop_collection(collection_name=self.collection_name)

    def col_info(self):
        """
        Get information about a collection.

        Returns:
            Dict[str, Any]: Collection information.
        """
        return self.client.get_collection_stats(collection_name=self.collection_name)

    def list(self, filters: dict = None, top_k: int = 100) -> list:
        """
        List all vectors in a collection.

        Args:
            filters (Dict, optional): Filters to apply to the list.
            top_k (int, optional): Number of vectors to return. Defaults to 100.

        Returns:
            List[OutputData]: List of vectors.
        """
        query_filter = self._create_filter(filters) if filters else None
        result = self.client.query(collection_name=self.collection_name, filter=query_filter, limit=top_k)
        memories = []
        for data in result:
            obj = OutputData(id=data.get("id"), score=None, payload=data.get("metadata"))
            memories.append(obj)
        return [memories]

    def reset(self):
        """Reset the index by deleting and recreating it."""
        logger.warning(f"Resetting index {self.collection_name}...")
        self.delete_col()
        self.create_col(self.collection_name, self.embedding_model_dims, self.metric_type)

File: mem0-ts/src/oss/src/vector_stores/milvus.ts
import { VectorStore } from "./base";
import { SearchFilters, VectorStoreConfig, VectorStoreResult } from "../types";

/**
 * Supported Milvus metric types. Mirrors the Python provider
 * (`mem0/configs/vector_stores/milvus.py`).
 */
export type MilvusMetricType = "L2" | "IP" | "COSINE" | "HAMMING" | "JACCARD";

export interface MilvusConfig extends VectorStoreConfig {
  /**
   * Full URL/address for the Milvus or Zilliz server.
   * Defaults to `http://localhost:19530`.
   */
  url?: string;
  /** Token / API key for Zilliz Cloud. Optional for a local setup. */
  token?: string;
  /** Name of the database. Optional (Milvus default database when empty). */
  dbName?: string;
  /** Collection name. Defaults to `mem0`. */
  collectionName?: string;
  /** Embedding dimensionality. Defaults to 1536 (OpenAI). */
  embeddingModelDims?: number;
  dimension?: number;
  /** Similarity metric. Defaults to `L2` (matches the Python provider). */
  metricType?: MilvusMetricType;
  /**
   * Pre-constructed `MilvusClient` instance. When provided, `url`/`token`/`dbName`
   * are ignored. Primarily useful for dependency injection in tests.
   */
  client?: any;
}

/**
 * Milvus vector store provider for the TypeScript OSS SDK.
 *
 * Mirrors the Python provider in `mem0/vector_stores/milvus.py`: dense-vector
 * CRUD (insert / search / get / update / delete / list + user-id helpers) plus
 * BM25 hybrid keyword search. New collections are created with a `text` +
 * `sparse` field pair and a BM25 function so `keywordSearch` can run full-text
 * search; collections created before BM25 support keep working with it disabled.
 *
 * The `@zilliz/milvus2-sdk-node` dependency is lazily required so the package
 * remains optional. Importing this module never forces the SDK to be installed
 * until a Milvus store is actually constructed.
 */
export class Milvus implements VectorStore {
  private client: any;
  private readonly collectionName: string;
  private readonly dimension: number;
  private readonly metricType: MilvusMetricType;
  private _initPromise?: Promise<void>;
  // Lazily-resolved SDK enums (set during client construction).
  private DataType: any;
  private FunctionType: any;
  // Whether this collection has the `text` + `sparse` fields for BM25 hybrid
  // search. Collections created before BM25 support lack them, so writing a
  // top-level `text` field or passing a sparse anns_field would be rejected.
  private hasBm25Schema = false;

  constructor(config: MilvusConfig) {
    this.collectionName = config.collectionName || "mem0";
    this.dimension = config.embeddingModelDims || config.dimension || 1536;
    this.metricType = config.metricType || "L2";

    if (config.client) {
      this.client = config.client;
      // Best-effort SDK enum resolution for an injected client (undefined when
      // the SDK isn't installed, e.g. unit tests that pass a fake client).
      try {
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        const sdk = require("@zilliz/milvus2-sdk-node");
        this.DataType = sdk.DataType;
        this.FunctionType = sdk.FunctionType;
      } catch (_) {
        this.DataType = undefined;
        this.FunctionType = undefined;
      }
    } else {
      let MilvusClient: any;
      let DataType: any;
      let FunctionType: any;
      try {
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        const sdk = require("@zilliz/milvus2-sdk-node");
        MilvusClient = sdk.MilvusClient;
        DataType = sdk.DataType;
        FunctionType = sdk.FunctionType;
      } catch (_) {
        throw new Error(
          "The '@zilliz/milvus2-sdk-node' package is required to use the Milvus vector store. " +
            "Install it with: npm install @zilliz/milvus2-sdk-node",
        );
      }
      this.DataType = DataType;
      this.FunctionType = FunctionType;
      this.client = new MilvusClient({
        address: config.url || "http://localhost:19530",
        token: config.token,
        database: config.dbName || undefined,
      });
    }

    this.initialize().catch((err) =>
      console.error("Error initializing Milvus:", err),
    );
  }

  async initialize(): Promise<void> {
    if (!this._initPromise) {
      this._initPromise = this.createCol(this.collectionName, this.dimension);
    }
    return this._initPromise;
  }

  /**
   * Create the collection if it does not already exist, with an AUTOINDEX
   * dense-vector index plus a BM25 `text` -> `sparse` full-text index. Idempotent
   * (mirrors the Python `create_col`). When the collection already exists, detect
   * whether the BM25 `text`/`sparse` fields are present so keyword search
   * degrades gracefully on collections created before BM25 support.
   */
  private async createCol(
    collectionName: string,
    vectorSize: number,
  ): Promise<void> {
    const has = await this.client.hasCollection({
      collection_name: collectionName,
    });
    // milvus2-sdk-node returns { value: boolean } for hasCollection.
    const exists = typeof has === "object" && has !== null ? has.value : has;
    if (exists) {
      // A pre-existing collection may predate BM25 support. Inspect its schema
      // so insert/search/keywordSearch know whether the text + sparse fields
      // exist instead of assuming and getting rejected by the server.
      const desc = await this.client.describeCollection({
        collection_name: collectionName,
      });
      const names = new Set(
        (desc?.schema?.fields || []).map((f: any) => f.name),
      );
      this.hasBm25Schema = names.has("text") && names.has("sparse");
      if (!this.hasBm25Schema) {
        console.warn(
          `Milvus collection '${collectionName}' predates BM25 hybrid search ` +
            "(no 'text'/'sparse' fields). Keyword scoring is disabled for it; " +
            "semantic search still works. Use a fresh collection to enable it.",
        );
      }
      return;
    }

    const DataType = this.DataType || {};
    const fields = [
      {
        name: "id",
        data_type: DataType.VarChar,
        is_primary_key: true,
        max_length: 512,
      },
      {
        name: "vectors",
        data_type: DataType.FloatVector,
        dim: vectorSize,
      },
      {
        name: "metadata",
        data_type: DataType.JSON,
      },
      // Analyzer-enabled text field that feeds the BM25 function below.
      {
        name: "text",
        data_type: DataType.VarChar,
        max_length: 65535,
        enable_analyzer: true,
      },
      // Sparse vectors are generated automatically by the BM25 function.
      {
        name: "sparse",
        data_type: DataType.SparseFloatVector,
      },
    ];

    await this.client.createCollection({
      collection_name: collectionName,
      fields,
      enable_dynamic_field: true,
      // BM25 turns the `text` field into `sparse` vectors for full-text search.
      functions: [
        {
          name: "bm25",
          type: this.FunctionType?.BM25,
          input_field_names: ["text"],
          output_field_names: ["sparse"],
          params: {},
        },
      ],
      index_params: [
        {
          field_name: "vectors",
          index_type: "AUTOINDEX",
          metric_type: this.metricType,
          index_name: "vector_index",
        },
        {
          field_name: "sparse",
          index_type: "SPARSE_INVERTED_INDEX",
          metric_type: "BM25",
          index_name: "sparse_index",
        },
      ],
    });

    await this.client.loadCollection({ collection_name: collectionName });
    this.hasBm25Schema = true;
  }

  /**
   * Filter keys are interpolated straight into the expression, so restrict them
   * to safe identifiers (same rule as the Python provider) to block injection.
   */
  private static readonly SAFE_FILTER_KEY = /^[a-zA-Z_][a-zA-Z0-9_]*$/;

  /**
   * Build a Milvus boolean filter expression from a flat filters object.
   * Mirrors the Python `_create_filter` (equality only, AND-combined): validate
   * each key, escape string values (backslash first, then double-quote), and
   * reject value types Milvus can't compare against a scalar field.
   */
  private createFilter(filters?: SearchFilters): string | undefined {
    if (!filters || Object.keys(filters).length === 0) return undefined;
    const operands: string[] = [];
    for (const [key, value] of Object.entries(filters)) {
      if (value === undefined || value === null) continue;
      if (!Milvus.SAFE_FILTER_KEY.test(key)) {
        throw new Error(`Invalid filter key: ${JSON.stringify(key)}`);
      }
      if (typeof value === "string") {
        // Escape backslashes before quotes so a value can't break out of the
        // string literal (order matters, exactly as in the Python provider).
        const escaped = value.replace(/\\/g, "\\\\").replace(/"/g, '\\"');
        operands.push(`(metadata["${key}"] == "${escaped}")`);
      } else if (typeof value === "number" || typeof value === "boolean") {
        operands.push(`(metadata["${key}"] == ${value})`);
      } else {
        throw new Error(
          `Filter value for ${JSON.stringify(key)} must be a string, number, or boolean, got ${typeof value}`,
        );
      }
    }
    return operands.length > 0 ? operands.join(" and ") : undefined;
  }

  /**
   * Text fed to the BM25 sparse index for a payload. Prefers the lemmatized
   * text, falls back to the raw memory `data`, and truncates to the VarChar
   * limit (mirrors the Python provider).
   */
  private bm25Text(payload?: Record<string, any>): string {
    if (!payload) return "";
    const raw = payload.text_lemmatized || payload.data || "";
    return String(raw).slice(0, 65535);
  }

  async insert(
    vectors: number[][],
    ids: string[],
    payloads: Record<string, any>[],
  ): Promise<void> {
    const data = vectors.map((vector, idx) => {
      const metadata = payloads[idx] || {};
      const row: Record<string, any> = {
        id: ids[idx],
        vectors: vector,
        metadata,
      };
      // Only write `text` when the collection has the BM25 schema; legacy
      // collections reject an unknown top-level field.
      if (this.hasBm25Schema) row.text = this.bm25Text(metadata);
      return row;
    });
    await this.client.insert({
      collection_name: this.collectionName,
      data,
    });
  }

  /**
   * Map raw Milvus search hits to VectorStoreResult. For the L2 metric,
   * distances are unbounded and smaller-is-better, so normalise them to a 0..1
   * similarity; every other metric passes the raw score through. Shared by
   * search and keywordSearch so both match the Python provider's `_parse_output`
   * (which likewise normalises by metric type regardless of dense vs BM25 score).
   */
  private parseHits(hits: any[]): VectorStoreResult[] {
    return hits.map((hit: any) => {
      const rawDistance = hit.score ?? hit.distance;
      let score = rawDistance;
      if (rawDistance != null && this.metricType === "L2") {
        score = 1.0 / (1.0 + rawDistance);
      }
      return {
        id: String(hit.id),
        payload: hit.metadata || {},
        score,
      } as VectorStoreResult;
    });
  }

  async search(
    query: number[],
    topK: number = 5,
    filters?: SearchFilters,
  ): Promise<VectorStoreResult[]> {
    const filter = this.createFilter(filters);
    const req: Record<string, any> = {
      collection_name: this.collectionName,
      data: [query],
      limit: topK,
      filter,
      output_fields: ["*"],
    };
    // A BM25 collection has both a dense `vectors` and a sparse `sparse` field,
    // so anns_field is ambiguous unless we name the dense one explicitly.
    if (this.hasBm25Schema) req.anns_field = "vectors";
    const res = await this.client.search(req);
    return this.parseHits(res?.results || []);
  }

  /**
   * BM25 full-text keyword search over the sparse field. Milvus tokenizes the
   * raw query string via the collection's BM25 function. Returns null when the
   * collection has no BM25 schema so callers fall back to dense search only
   * (mirrors the Python provider).
   */
  async keywordSearch(
    query: string,
    topK: number = 5,
    filters?: SearchFilters,
  ): Promise<VectorStoreResult[] | null> {
    if (!this.hasBm25Schema) return null;
    try {
      const filter = this.createFilter(filters);
      const res = await this.client.search({
        collection_name: this.collectionName,
        data: [query],
        anns_field: "sparse",
        limit: topK,
        filter,
        output_fields: ["*"],
      });
      return this.parseHits(res?.results || []);
    } catch (_) {
      // Keyword search is best-effort; degrade to null rather than failing the
      // whole retrieval path.
      return null;
    }
  }

  async get(vectorId: string): Promise<VectorStoreResult | null> {
    const res = await this.client.get({
      collection_name: this.collectionName,
      ids: [vectorId],
      output_fields: ["id", "metadata"],
    });
    const rows = res?.data || [];
    if (!rows.length) return null;
    return {
      id: String(rows[0].id),
      payload: rows[0].metadata || {},
    };
  }

  async update(
    vectorId: string,
    vector: number[],
    payload: Record<string, any>,
  ): Promise<void> {
    const row: Record<string, any> = {
      id: vectorId,
      vectors: vector,
      metadata: payload,
    };
    if (this.hasBm25Schema) row.text = this.bm25Text(payload);
    await this.client.upsert({
      collection_name: this.collectionName,
      data: [row],
    });
  }

  async delete(vectorId: string): Promise<void> {
    await this.client.delete({
      collection_name: this.collectionName,
      ids: [vectorId],
    });
  }

  async deleteCol(): Promise<void> {
    await this.client.dropCollection({
      collection_name: this.collectionName,
    });
  }

  async list(
    filters?: SearchFilters,
    topK: number = 100,
  ): Promise<[VectorStoreResult[], number]> {
    const filter = this.createFilter(filters);
    const res = await this.client.query({
      collection_name: this.collectionName,
      filter: filter ?? "",
      limit: topK,
      output_fields: ["id", "metadata"],
    });
    const rows = res?.data || [];
    const results: VectorStoreResult[] = rows.map((row: any) => ({
      id: String(row.id),
      payload: row.metadata || {},
    }));
    return [results, results.length];
  }

  private generateUUID(): string {
    return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
      const r = (Math.random() * 16) | 0;
      const v = c === "x" ? r : (r & 0x3) | 0x8;
      return v.toString(16);
    });
  }

  private async ensureMigrationsCol(): Promise<void> {
    const name = "memory_migrations";
    const has = await this.client.hasCollection({ collection_name: name });
    const exists = typeof has === "object" && has !== null ? has.value : has;
    if (exists) return;

    const DataType = this.DataType || {};
    await this.client.createCollection({
      collection_name: name,
      fields: [
        {
          name: "id",
          data_type: DataType.VarChar,
          is_primary_key: true,
          max_length: 512,
        },
        // Milvus rejects a FloatVector with dim < 2, so use the minimum (2). This
        // helper collection is never vector-searched; the vector is a fixed
        // placeholder that exists only to satisfy the required-vector-field schema.
        { name: "vectors", data_type: DataType.FloatVector, dim: 2 },
        { name: "user_id", data_type: DataType.VarChar, max_length: 512 },
      ],
      index_params: [
        {
          field_name: "vectors",
          index_type: "AUTOINDEX",
          // Fixed metric: this helper collection is never vector-searched, and a
          // zero vector has no direction, so COSINE would be degenerate here.
          metric_type: "L2",
          index_name: "vector_index",
        },
      ],
    });
    await this.client.loadCollection({ collection_name: name });
  }

  async getUserId(): Promise<string> {
    await this.ensureMigrationsCol();
    const res = await this.client.query({
      collection_name: "memory_migrations",
      filter: "",
      limit: 1,
      output_fields: ["*"],
    });
    const rows = res?.data || [];
    if (rows.length > 0 && rows[0].user_id) {
      return String(rows[0].user_id);
    }
    const randomUserId =
      Math.random().toString(36).substring(2, 15) +
      Math.random().toString(36).substring(2, 15);
    await this.client.insert({
      collection_name: "memory_migrations",
      data: [
        { id: this.generateUUID(), vectors: [0, 0], user_id: randomUserId },
      ],
    });
    return randomUserId;
  }

  async setUserId(userId: string): Promise<void> {
    await this.ensureMigrationsCol();
    // Keep a single row: reuse the existing row's id when present so the upsert
    // overwrites it in place instead of appending a new row on every call
    // (mirrors the qdrant provider's single-row telemetry id).
    const existing = await this.client.query({
      collection_name: "memory_migrations",
      filter: "",
      limit: 1,
      output_fields: ["id"],
    });
    const rows = existing?.data || [];
    const id =
      rows.length > 0 && rows[0].id ? String(rows[0].id) : this.generateUUID();
    await this.client.upsert({
      collection_name: "memory_migrations",
      data: [{ id, vectors: [0, 0], user_id: userId }],
    });
  }
}
